src/pool.rs

```diff
@@ -140,21 +145,23 @@ impl ConnectionPool {
             let len = locked.len.load(Ordering::SeqCst);
             if len < locked.size {
                 let socket = try!(self.connect());
-                let _ = locked.len.fetch_add(1, Ordering::SeqCst);
-                return Ok(PooledStream {
+                let mut stream = PooledStream {
                     socket: Some(socket),
                     pool: self.inner.clone(),
                     wait_lock: self.wait_lock.clone(),
                     iteration: locked.iteration,
-                });
+                };
+
+                self.handshake(client, &mut stream)?;
+                let _ = locked.len.fetch_add(1, Ordering::SeqCst);
```

Comment:
I thought about this more and realized an interesting issue here — the PooledStream is returned to the pool via the `Drop` trait, which means that if the handshake fails, 
- `acquire_stream` returns an error instead of the stream
- the stream is dropped
- the new socket is added into the pool of available connections anyways

If we're okay with the possibility of a socket being opened _without_ a successful handshake, we should increment this _after_ the PooledStream is created, but before the handshake.

If we're not okay with that possibility, we need to do extra work — either add a "completed_handshake" kinda boolean flag to pooled streams, or figure out something clever where we don't have to create the PooledStream until we handshake.

Reply:
Another good catch! I went with the boolean flag on the stream; I don't think the overhead of a few extra booleans should make a noticeable difference, and I think it's important to make sure that every connection handshakes as the spec requires. (I also didn't rebase this time so that the changes are easier to follow; I'll just rebase to clean things up before merging once we're done making changes)